man/ejabberd.yml.5
'\" t
.\"     Title: ejabberd.yml
.\"    Author: [see the "AUTHOR" section]
.\" Generator: DocBook XSL Stylesheets vsnapshot <http://docbook.sf.net/>
.\"      Date: 09/29/2026
.\"    Manual: \ \&
.\"    Source: \ \&
.\"  Language: English
.\"
.TH "EJABBERD\&.YML" "5" "09/29/2026" "\ \&" "\ \&"
.\" -----------------------------------------------------------------
.\" * Define some portability stuff
.\" -----------------------------------------------------------------
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" http://bugs.debian.org/507673
.\" http://lists.gnu.org/archive/html/groff/2009-02/msg00013.html
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\" -----------------------------------------------------------------
.\" * set default formatting
.\" -----------------------------------------------------------------
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.\" -----------------------------------------------------------------
.\" * MAIN CONTENT STARTS HERE *
.\" -----------------------------------------------------------------
.SH "NAME"
ejabberd.yml \- main configuration file for ejabberd\&.
.SH "SYNOPSIS"
.sp
ejabberd\&.yml
.SH "DESCRIPTION"
.sp
The configuration file is written in YAML language\&.
.if n \{\
.sp
.\}
.RS 4
.it 1 an-trap
.nr an-no-space-flag 1
.nr an-break-flag 1
.br
.ps +1
\fBWarning\fR
.ps -1
.br
.sp
YAML is indentation sensitive, so make sure you respect indentation, or otherwise you will get pretty cryptic configuration errors\&.
.sp .5v
.RE
.sp
Logically, configuration options are split into 3 main categories: \fIModules\fR, \fIListeners\fR and everything else called \fITop Level\fR options\&. Thus this document is split into 3 main chapters describing each category separately\&. So, the contents of ejabberd\&.yml will typically look like this:
.sp
.if n \{\
.RS 4
.\}
.nf
hosts:
  \- example\&.com
  \- domain\&.tld
loglevel: info
\&.\&.\&.
listen:
  \-
    port: 5222
    module: ejabberd_c2s
  \&.\&.\&.
modules:
  mod_roster: {}
  \&.\&.\&.
.fi
.if n \{\
.RE
.\}
.sp
Any configuration error (such as syntax error, unknown option or invalid option value) is fatal in the sense that ejabberd will refuse to load the whole configuration file and will not start or will abort configuration reload\&.
.sp
All options can be changed in runtime by running \fIejabberdctl reload\-config\fR command\&. Configuration reload is atomic: either all options are accepted and applied simultaneously or the new configuration is refused without any impact on currently running configuration\&.
.sp
Some options can be specified for particular virtual host(s) only using \fIhost_config\fR or \fIappend_host_config\fR options\&. Such options are called \fIlocal\fR\&. Examples are \fImodules\fR, \fIauth_method\fR and \fIdefault_db\fR\&. The options that cannot be defined per virtual host are called \fIglobal\fR\&. Examples are \fIloglevel\fR, \fIcertfiles\fR and \fIlisten\fR\&. It is a configuration mistake to put \fIglobal\fR options under \fIhost_config\fR or \fIappend_host_config\fR section \- ejabberd will refuse to load such configuration\&.
.sp
It is not recommended to write ejabberd\&.yml from scratch\&. Instead it is better to start from "default" configuration file available at https://github\&.com/processone/ejabberd/blob/26\&.09/ejabberd\&.yml\&.example\&. Once you get ejabberd running you can start changing configuration options to meet your requirements\&.
.sp
Note that this document is intended to provide comprehensive description of all configuration options that can be consulted to understand the meaning of a particular option, its format and possible values\&. It will be quite hard to understand how to configure ejabberd by reading this document only \- for this purpose the reader is recommended to read online Configuration Guide available at https://docs\&.ejabberd\&.im/admin/configuration\&.
.SH "TOP LEVEL OPTIONS"
.sp
This section describes top level options of ejabberd 26\&.09\&. The options that changed in this version are marked with 🟠\&.
.PP
\fBaccess_rules\fR: \fI{AccessName: {allow|deny: ACLName|ACLDefinition}}\fR
.RS 4
This option defines
\fIbasic\&.md#access\-rules|Access Rules\fR\&. Each access rule is assigned a name that can be referenced from other parts of the configuration file (mostly from
\fIaccess\fR
options of ejabberd modules)\&. Each rule definition may contain arbitrary number of
\fIallow\fR
or
\fIdeny\fR
sections, and each section may contain any number of ACL rules (see
\fIacl\fR
option)\&. There are no access rules defined by default\&.
.sp
\fBExample\fR:
.sp
.if n \{\
.RS 4
.\}
.nf
access_rules:
  configure:
    allow: admin
  something:
    deny: someone
    allow: all
  s2s_banned:
    deny: problematic_hosts
    deny: banned_forever
    deny:
      ip: 222\&.111\&.222\&.111/32
    deny:
      ip: 111\&.222\&.111\&.222/32
    allow: all
  xmlrpc_access:
    allow:
      user: peter@example\&.com
    allow:
      user: ivone@example\&.com
    allow:
      user: bot@example\&.com
      ip: 10\&.0\&.0\&.0/24
.fi
.if n \{\
.RE
.\}
.RE
.PP
\fBacl\fR: \fI{ACLName: {ACLType: ACLValue}}\fR
.RS 4
This option defines
\fI\&.\&./configuration/basic\&.md#acl|access control lists\fR: named sets of rules which are used to match against different targets (such as a JID or an IP address)\&. Every set of rules has name
\fIACLName\fR: it can be any string except
\fIall\fR
or
\fInone\fR
(those are predefined names for the rules that match all or nothing respectively)\&. The name
\fIACLName\fR
can be referenced from other parts of the configuration file, for example in
\fIaccess_rules\fR
option\&. The rules of
\fIACLName\fR
are represented by mapping
\fI{ACLType: ACLValue}\fR\&. These can be one of the following:
.PP
\fBip\fR: \fINetwork\fR
.RS 4
The rule matches any IP address from the
\fINetwork\fR\&.
.RE
.PP
\fBnode_glob\fR: \fIPattern\fR
.RS 4
Same as
\fInode_regexp\fR, but matching is performed on a specified
\fIPattern\fR
according to the rules used by the Unix shell\&.
.RE
.PP
\fBnode_regexp\fR: \fIuser_regexp@server_regexp\fR
.RS 4
The rule matches any JID with node part matching regular expression
\fIuser_regexp\fR
and server part matching regular expression
\fIserver_regexp\fR\&.
.RE
.PP
\fBresource\fR: \fIResource\fR
.RS 4
The rule matches any JID with a resource
\fIResource\fR\&.
.RE
.PP
\fBresource_glob\fR: \fIPattern\fR
.RS 4
Same as
\fIresource_regexp\fR, but matching is performed on a specified
\fIPattern\fR
according to the rules used by the Unix shell\&.
.RE
.PP
\fBresource_regexp\fR: \fIRegexp\fR
.RS 4
The rule matches any JID with a resource that matches regular expression
\fIRegexp\fR\&.
.RE
.PP
\fBserver\fR: \fIServer\fR
.RS 4
The rule matches any JID from server
\fIServer\fR\&. The value of
\fIServer\fR
must be a valid hostname or an IP address\&.
.RE
.PP
\fBserver_glob\fR: \fIPattern\fR
.RS 4
Same as
\fIserver_regexp\fR, but matching is performed on a specified
\fIPattern\fR
according to the rules used by the Unix shell\&.
.RE
.PP
\fBserver_regexp\fR: \fIRegexp\fR
.RS 4
The rule matches any JID from the server that matches regular expression
\fIRegexp\fR\&.
.RE
.PP
\fBuser\fR: \fIUsername\fR
.RS 4
If
\fIUsername\fR
is in the form of "user@server", the rule matches a JID against this value\&. Otherwise, if
\fIUsername\fR
is in the form of "user", the rule matches any JID that has
\fIUsername\fR
in the node part as long as the server part of this JID is any virtual host served by ejabberd\&.
.RE
.PP
\fBuser_glob\fR: \fIPattern\fR
.RS 4
Same as
\fIuser_regexp\fR, but matching is performed on a specified
\fIPattern\fR
according to the rules used by the Unix shell\&.
.RE
.PP
\fBuser_regexp\fR: \fIRegexp\fR
.RS 4
If
\fIRegexp\fR
is in the form of "regexp@server", the rule matches any JID with node part matching regular expression "regexp" as long as the server part of this JID is equal to "server"\&. If
\fIRegexp\fR
is in the form of "regexp", the rule matches any JID with node part matching regular expression "regexp" as long as the server part of this JID is any virtual host served by ejabberd\&.
.RE
.RE
.PP
\fBacme\fR: \fIOptions\fR
.RS 4
\fIbasic\&.md#acme|ACME\fR
configuration, to automatically obtain SSL certificates for the domains served by ejabberd, which means that certificate requests and renewals are performed to some CA server (aka "ACME server") in a fully automated mode\&. The
\fIOptions\fR
are:
.PP
\fBauto\fR: \fItrue | false\fR
.RS 4
Whether to automatically request certificates for all configured domains (that yet have no a certificate) on server start or configuration reload\&. The default is
\fItrue\fR\&.
.RE
.PP
\fBca_url\fR: \fIURL\fR
.RS 4
The ACME directory URL used as an entry point for the ACME server\&. The default value is
https://acme\-v02\&.api\&.letsencrypt\&.org/directory
\- the directory URL of Let\(cqs Encrypt authority\&.
.RE
.PP
\fBcert_type\fR: \fIrsa | ec\fR
.RS 4
A type of a certificate key\&. Available values are
\fIec\fR
and
\fIrsa\fR
for EC and RSA certificates respectively\&. It\(cqs better to have RSA certificates for the purpose of backward compatibility with legacy clients and servers, thus the default is
\fIrsa\fR\&.
.RE
.PP
\fBcontact\fR: \fI[Contact, \&.\&.\&.]\fR
.RS 4
A list of contact addresses (typically emails) where an ACME server will send notifications when problems occur\&. The value of
\fIContact\fR
must be in the form of "scheme:address" (e\&.g\&. "mailto:user@domain\&.tld")\&. The default is an empty list which means an ACME server will send no notices\&.
.RE
.sp
\fBExample\fR:
.sp
.if n \{\
.RS 4
.\}
.nf
acme:
  ca_url: https://acme\-v02\&.api\&.letsencrypt\&.org/directory
  contact:
    \- mailto:admin@domain\&.tld
    \- mailto:bot@domain\&.tld
  auto: true
  cert_type: rsa
.fi
.if n \{\
.RE
.\}
.RE
.PP
\fBallow_contrib_modules\fR: \fItrue | false\fR
.RS 4
Whether to allow installation of third\-party modules or not\&. See
\fI\&.\&./\&.\&./admin/guide/modules\&.md#ejabberd\-modules|ejabberd\-modules\fR
documentation section\&. The default value is
\fItrue\fR\&.
.RE
.PP
\fBallow_multiple_connections\fR: \fItrue | false\fR
.RS 4
This option is only used when the anonymous mode is enabled\&. Setting it to
\fItrue\fR
means that the same username can be taken multiple times in anonymous login mode if different resource are used to connect\&. This option is only useful in very special occasions\&. The default value is
\fIfalse\fR\&.
.RE
.PP
\fBanonymous_protocol\fR: \fIlogin_anon | sasl_anon | both\fR
.RS 4
Define what
\fIauthentication\&.md#anonymous\-login\-and\-sasl\-anonymous|anonymous\fR
protocol will be used:
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
\fIlogin_anon\fR
means that the anonymous login method will be used\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
\fIsasl_anon\fR
means that the SASL Anonymous method will be used\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
\fIboth\fR
means that SASL Anonymous and login anonymous are both enabled\&.
.RE
.RE
.sp
The default value is \fIsasl_anon\fR\&.
.PP
\fBapi_permissions\fR: \fI[Permission, \&.\&.\&.]\fR
.RS 4
Define the permissions for API access\&. Please consult the ejabberd Docs web → For Developers → ejabberd ReST API →
\fI\&.\&./\&.\&./developer/ejabberd\-api/permissions\&.md|API Permissions\fR\&.
.RE
.PP
\fBappend_host_config\fR: \fI{Host: Options}\fR
.RS 4
Add a few specific options to a certain
\fI\&.\&./configuration/basic\&.md#virtual\-hosting|virtual host\fR\&.
.RE
.PP
\fBauth_cache_life_time\fR: \fItimeout()\fR
.RS 4
Same as
\fIcache_life_time\fR, but applied to authentication cache only\&. If not set, the value from
\fIcache_life_time\fR
will be used\&.
.RE
.PP
\fBauth_cache_missed\fR: \fItrue | false\fR
.RS 4
Same as
\fIcache_missed\fR, but applied to authentication cache only\&. If not set, the value from
\fIcache_missed\fR
will be used\&.
.RE
.PP
\fBauth_cache_size\fR: \fIpos_integer() | infinity\fR
.RS 4
Same as
\fIcache_size\fR, but applied to authentication cache only\&. If not set, the value from
\fIcache_size\fR
will be used\&.
.RE
.PP
\fBauth_external_user_exists_check 🟠\fR: \fItrue | false | force\fR
.RS 4
\fINote\fR
about this option: added in 23\&.10,
force
added in 26\&.09\&. Supplement check for user existence based on
\fImod_last\fR
data, for authentication methods that don\(cqt have a way to reliably tell if a user exists (like is the case for
\fIjwt\fR
and certificate based authentication)\&. This helps with processing offline message for those users\&. Using
force
values enables this processing even when no external storage method is enabled (like when using certificate authentications together with local stored passwords\&. The default value is
\fItrue\fR\&.
.RE
.PP
\fBauth_method\fR: \fI[mnesia | sql | anonymous | external | jwt | ldap | pam, \&.\&.\&.]\fR
.RS 4
A list of
\fIauthentication\&.md|authentication\fR
methods to use\&. If several methods are defined, authentication is considered successful as long as authentication of at least one of the methods succeeds\&. The default value is
\fI[mnesia]\fR\&.
.RE
.PP
\fBauth_opts\fR: \fI[Option, \&.\&.\&.]\fR
.RS 4
This is used by the contributed module
\fIejabberd_auth_http\fR
that can be installed from the
\fI\&.\&./\&.\&./admin/guide/modules\&.md#ejabberd\-contrib|ejabberd\-contrib\fR
Git repository\&. Please refer to that module\(cqs README file for details\&.
.RE
.PP
\fBauth_password_format\fR: \fIplain | scram\fR
.RS 4
\fINote\fR
about this option: improved in 20\&.01\&. The option defines in what format the users passwords are stored, plain text or in
\fIauthentication\&.md#scram|SCRAM\fR
format:
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
\fIplain\fR: The password is stored as plain text in the database\&. This is risky because the passwords can be read if your database gets compromised\&. This is the default value\&. This format allows clients to authenticate using: the old Jabber Non\-SASL (XEP\-0078), SASL PLAIN, SASL DIGEST\-MD5, and SASL SCRAM\-SHA\-1/256/512(\-PLUS)\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
\fIscram\fR: The password is not stored, only some information required to verify the hash provided by the client\&. It is impossible to obtain the original plain password from the stored information; for this reason, when this value is configured it cannot be changed to plain anymore\&. This format allows clients to authenticate using: SASL PLAIN and SASL SCRAM\-SHA\-1/256/512(\-PLUS)\&. The SCRAM variant depends on the
\fIauth_scram_hash\fR
option\&.
.RE
.RE
.sp
The default value is \fIplain\fR\&.
.PP
\fBauth_password_types_hidden_in_sasl1\fR: \fI[plain | scram_sha1 | scram_sha256 | scram_sha512]\fR
.RS 4
\fINote\fR
about this option: added in 25\&.07\&. List of password types that should not be offered in SASL1 authenticatication\&. Because SASL1, unlike SASL2, can\(cqt have list of available mechanisms tailored to individual user, it\(cqs possible that offered mechanisms will not be compatible with stored password, especially if new password type was added recently\&. This option allows disabling offering some mechanisms in SASL1, to a time until new password type will be available for all users\&.
.RE
.PP
\fBauth_scram_hash\fR: \fIsha | sha256 | sha512\fR
.RS 4
Hash algorithm that should be used to store password in
\fIauthentication\&.md#scram|SCRAM\fR
format\&. You shouldn\(cqt change this if you already have passwords generated with a different algorithm \- users that have such passwords will not be able to authenticate\&. The default value is
\fIsha\fR\&.
.RE
.PP
\fBauth_stored_password_types\fR: \fI[plain | scram_sha1 | scram_sha256 | scram_sha512]\fR
.RS 4
\fINote\fR
about this option: added in 25\&.03\&. List of password types that should be stored simultaneously for each user in database\&. When the user sets the account password, database will be updated to store the password in formats compatible with each type listed here\&. This can be used to migrate user passwords to a more secure format\&. If this option if set, it will override values set in
\fIauth_scram_hash\fR
and
\fIauth_password_format\fR
options\&. The default value is
[]\&.
.RE
.PP
\fBauth_use_cache\fR: \fItrue | false\fR
.RS 4
Same as
\fIuse_cache\fR, but applied to authentication cache only\&. If not set, the value from
\fIuse_cache\fR
will be used\&.
.RE
.PP
\fBc2s_cafile\fR: \fIPath\fR
.RS 4
Full path to a file containing one or more CA certificates in PEM format\&. All client certificates should be signed by one of these root CA certificates and should contain the corresponding JID(s) in
\fIsubjectAltName\fR
field\&. There is no default value\&.
.RE
.sp
You can use \fIhost_config\fR to specify this option per\-vhost\&.
.sp
To set a specific file per listener, use the listener\(cqs \fIlisten\-options\&.md#cafile|cafile\fR option\&. Please notice that \fIc2s_cafile\fR overrides the listener\(cqs \fIcafile\fR option\&.
.PP
\fBc2s_ciphers\fR: \fI[Cipher, \&.\&.\&.]\fR
.RS 4
A list of OpenSSL ciphers to use for c2s connections\&. The default value is shown in the example below:
.sp
\fBExample\fR:
.sp
.if n \{\
.RS 4
.\}
.nf
c2s_ciphers:
  \- HIGH
  \- "!aNULL"
  \- "!eNULL"
  \- "!3DES"
  \- "@STRENGTH"
.fi
.if n \{\
.RE
.\}
.RE
.PP
\fBc2s_dhfile\fR: \fIPath\fR
.RS 4
Full path to a file containing custom DH parameters to use for c2s connections\&. Such a file could be created with the command

ChangeLog
2009-05-12  Christophe Romain <christophe.romain@process-one.net>

	* src/mod_pubsub/mod_pubsub.erl: Fix bad transaction returncode in
	unsubscribe presence handler

2009-05-11  Christophe Romain <christophe.romain@process-one.net>

	* src/mod_caps.erl: Prevent race condition when calling get_caps while
	note_caps has not been handled yet (EJAB-934)
	* src/mod_pubsub/mod_pubsub.erl: Likewise

2009-05-08  Christophe Romain <christophe.romain@process-one.net>

	* src/mod_pubsub/mod_pubsub.erl: Allow to get subscriptions on a given
	node (EJAB-712); Improve XML schema validation on publish item;
	Remove subscriptions of users whose authorization has been removed on
	nodes with presence access model; Do not allow unsubscribed user to
	subscribe node with presence access model.

	* src/mod_pubsub/mod_pubsub.erl: Handle Example 90 of XEP-0060
	(EJAB-909)

2009-05-08  Mickael Remond  <mremond@process-one.net>

	* src/ejabberd_app.erl: Better support for profiling.
	* src/ejabberd.erl: Likewise
	* src/ejabberd_ctl.erl: Likewise
	* src/ejabberd_debug.erl: Likewise

2009-05-07  Christophe Romain <christophe.romain@process-one.net>

	* src/mod_caps.erl: Set debug message to DEBUG (from debian patch,
	thanks to Sergei Golovan)

	* src/mod_pubsub/mod_pubsub.erl: Remove subscriptions when anonymous
	user removed (EJAB-913) (thanks to Andy Skelton)

	* src/mod_pubsub/mod_pubsub.erl: Fix disco#items bug on root node,
	and fix other minor typo from previous patch.

	* src/mod_pubsub/mod_pubsub.erl: Avoid calling get_user_resources
	on non local domain when pep_sendlast_offline is enabled

	* src/mod_pubsub/mod_pubsub.erl: Reduce send_last_item load and number
	of calls to get_caps

	* src/mod_pubsub/mod_pubsub.erl: Fix get_entity_* not returning node
	* src/mod_pubsub/node_default.erl: Likewise
	* src/mod_pubsub/nodetree_default.erl: Likewise

	* src/mod_pubsub/mod_pubsub.erl: Retract policy should obey
	pubsub#publish_model (EJAB-871) (thanks to Matthew Baron)
	* src/mod_pubsub/node_default.erl: Likewise
	* src/mod_pubsub/node_mb.erl: Likewise
	* src/mod_pubsub/node_dispatch.erl: Likewise
	* src/mod_pubsub/node_buddy.erl: Likewise
	* src/mod_pubsub/node_private.erl: Likewise
	* src/mod_pubsub/node_public.erl: Likewise
	* src/mod_pubsub/node_default.erl: Likewise
	* src/mod_pubsub/node_pep.erl: Likewise
	* src/mod_pubsub/node_club.erl: Likewise
	* src/mod_pubsub/node_flat.erl: Likewise
	* src/mod_pubsub/node.template: Likewise

2009-05-06  Badlop  <badlop@process-one.net>

	* src/ejabberd_c2s.erl: Replace TYPE/1 with is_TYPE/1 (EJAB-922)
	* src/ejabberd_logger_h.erl: 
	* src/ejabberd_s2s.erl: 
	* src/eldap/eldap.erl: 
	* src/mod_offline.erl: 
	* src/mod_offline_odbc.erl: 
	* src/mod_roster.erl: 
	* src/odbc/odbc_queries.erl: 
	* src/p1_fsm.erl: 
	* src/xml.erl: 

	* src/Makefile.in: Prevent Erlang R13B compilation warning:
	behaviour X undefined (EJAB-920)

	* src/ejabberd_loglevel.erl: Use dynamic_compile instead of
	ram_file_io_server. Support definition of loglevels with integer
	or atom. (thanks to Geoff Cant)(EJAB-919)
	* src/dynamic_compile.erl: Added erlang module that converts
	string to binary loadable code by Mats Cronqvist, Chris Newcombe,
	and Jacob Vorreuter.
	* src/ram_file_io_server.erl: Remove file not longer useful.
	* src/ejabberd.app: Likewise

2009-05-03  Badlop  <badlop@process-one.net>

	* src/mod_muc/mod_muc_room.erl: Fix badarg return (EJAB-899)

2009-04-30  Christophe Romain <christophe.romain@process-one.net>

	* src/mod_pubsub/mod_pubsub.erl: API change for major optimization
	* src/mod_pubsub/pubsub.hrl: Likewise
	* src/mod_pubsub/nodetree_default.erl: Likewise
	* src/mod_pubsub/nodetree_virtual.erl: Likewise
	* src/mod_pubsub/node_mb.erl: Likewise
	* src/mod_pubsub/node_dispatch.erl: Likewise
	* src/mod_pubsub/node_buddy.erl: Likewise
	* src/mod_pubsub/node_private.erl: Likewise
	* src/mod_pubsub/node_public.erl: Likewise
	* src/mod_pubsub/node_default.erl: Likewise
	* src/mod_pubsub/node_pep.erl: Likewise
	* src/mod_pubsub/node_club.erl: Likewise
	* src/mod_pubsub/node_flat.erl: Likewise
	* src/mod_pubsub/node.template: Likewise
	* src/mod_pubsub/gen_pubsub_node.erl: Likewise
	* src/mod_pubsub/gen_pubsub_nodetree.erl: Likewise

	* src/mod_caps.erl: Reduce memory consumption and remove mnesia table

2009-04-28  Badlop  <badlop@process-one.net>

	* src/ejabberd_hooks.erl: Support distributed hooks (EJAB-829)

	* src/ejabberd_hooks.erl: anonymous functions support.

2009-04-27  Badlop  <badlop@process-one.net>

	* src/translate.erl: Support additional files (EJAB-925)
	* contrib/extract_translations/extract_translations.erl: Likewise
	* contrib/extract_translations/prepare-translation.sh: Likewise

	* src/win32_dns.erl: Fix problem parsing some win32 dns (EJAB-927)

	* src/cyrsasl_digest.erl: Fix auth verification (EJAB-863)

	* src/jlib.erl: Fix recursive call to speedup base64
	decoding (thanks to Jeffrey Rogiers)(EJAB-333)

2009-04-24  Christophe Romain <christophe.romain@process-one.net>

	* src/odbc/ejabberd_odbc.erl: allow to run query bloc as erlang
	function without transaction

	* src/mod_pubsub/mod_pubsub.erl: do not register handlers and hooks
	while plugins and ets tables are not initialized.

2009-04-23  Christophe Romain <christophe.romain@process-one.net>

	* src/mod_pubsub/mod_pubsub.erl: improve send last published items
	(not spawned as much) and allow to send last PEP items of our offline
	contacts if configured for (fix discussion issue on standars ML)

2009-04-22  Badlop  <badlop@process-one.net>

	* src/ejabberd.cfg.example: Fix English typos. Fix line length:
	max 80 characters per line. Remove trailing blankspaces. Added
	markers for Vim fold, you may want to add to $HOME/.vimrc this
	line: set modeline
	* src/ejabberdctl.cfg.example: Likewise

	* src/mod_muc/mod_muc.erl: Limit number of characters in Room ID,
	Name and Description (EJAB-899)
	* src/mod_muc/mod_muc_room.erl: Likewise
	* doc/guide.tex: Likewise
	* doc/guide.html: Likewise


src/ejabberd.erl
%%%----------------------------------------------------------------------
%%% File    : ejabberd.erl
%%% Author  : Alexey Shchepin <alexey@process-one.net>
%%% Purpose : ejabberd wrapper: start / stop
%%% Created : 16 Nov 2002 by Alexey Shchepin <alexey@process-one.net>
%%%
%%%
%%% ejabberd, Copyright (C) 2002-2026   ProcessOne
%%%
%%% This program is free software; you can redistribute it and/or
%%% modify it under the terms of the GNU General Public License as
%%% published by the Free Software Foundation; either version 2 of the
%%% License, or (at your option) any later version.
%%%
%%% This program is distributed in the hope that it will be useful,
%%% but WITHOUT ANY WARRANTY; without even the implied warranty of
%%% MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
%%% General Public License for more details.
%%%
%%% You should have received a copy of the GNU General Public License along
%%% with this program; if not, write to the Free Software Foundation, Inc.,
%%% 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
%%%
%%%----------------------------------------------------------------------

-module(ejabberd).
-author('alexey@process-one.net').
-compile({no_auto_import, [{halt, 0}]}).

-protocol({rfc, 6122}).
-protocol({rfc, 7590}).
-protocol({xep, 4, '2.13.2', '0.5.0', "complete", ""}).
-protocol({xep, 59, '1.0', '2.1.0', "complete", ""}).
-protocol({xep, 82, '1.1.1', '2.1.0', "complete", ""}).
-protocol({xep, 86, '1.0', '0.5.0', "complete", ""}).
-protocol({xep, 106, '1.1.1', '0.5.0', "complete", ""}).
-protocol({xep, 170, '1.0', '17.12', "complete", ""}).
-protocol({xep, 178, '1.2', '17.03', "complete", ""}).
-protocol({xep, 205, '1.0.2', '1.1.2', "complete", ""}).
-protocol({xep, 368, '1.1.0', '17.09', "complete", ""}).
-protocol({xep, 386, '0.3.0', '24.02', "complete", ""}).
-protocol({xep, 388, '1.0.4', '24.02', "complete", ""}).
-protocol({xep, 440, '1.0.0', '24.02', "complete", ""}).
-protocol({xep, 474, '0.5.0', '24.02', "complete", "0.4.0 since 25.03"}).

-export([start/0, stop/0, halt/0, start_app/1, start_app/2,
	 get_pid_file/0, check_apps/0, module_name/1, is_loaded/0]).

-include("logger.hrl").

start() ->
    case application:ensure_all_started(ejabberd) of
      {error, Err} -> error_logger:error_msg("Failed to start ejabberd application: ~p", [Err]);
      Ok -> Ok
    end.

stop() ->
    application:stop(ejabberd).

halt() ->
    ejabberd_logger:flush(),
    erlang:halt(1, [{flush, true}]).

-spec get_pid_file() -> false | string().
get_pid_file() ->
    case os:getenv("EJABBERD_PID_PATH") of
	false ->
	    false;
	"" ->
	    false;
	Path ->
	    Path
    end.

start_app(App) ->
    start_app(App, temporary).

start_app(App, Type) ->
    StartFlag = not is_loaded(),
    start_app(App, Type, StartFlag).

is_loaded() ->
    Apps = application:which_applications(),
    lists:keymember(ejabberd, 1, Apps).

start_app(App, Type, StartFlag) when is_atom(App) ->
    start_app([App], Type, StartFlag);
start_app([App|Apps], Type, StartFlag) ->
    case application:start(App,Type) of
        ok ->
            start_app(Apps, Type, StartFlag);
        {error, {already_started, _}} ->
            start_app(Apps, Type, StartFlag);
        {error, {not_started, DepApp}} ->
            case lists:member(DepApp, [App|Apps]) of
                true ->
                    Reason = io_lib:format(
                               "Failed to start Erlang application '~ts': "
                               "circular dependency with '~ts' detected",
                               [App, DepApp]),
                    exit_or_halt(Reason, StartFlag);
                false ->
                    start_app([DepApp,App|Apps], Type, StartFlag)
            end;
        {error, Why} ->
            Reason = io_lib:format(
		       "Failed to start Erlang application '~ts': ~ts. ~ts",
		       [App, format_error(Why), hint()]),
            exit_or_halt(Reason, StartFlag)
    end;
start_app([], _Type, _StartFlag) ->
    ok.

check_app_modules(App, StartFlag) ->
    case application:get_key(App, modules) of
        {ok, Mods} ->
            lists:foreach(
              fun(Mod) ->
                      case code:which(Mod) of
                          non_existing ->
                              File = get_module_file(App, Mod),
                              Reason = io_lib:format(
                                         "Couldn't find file ~ts needed "
					 "for Erlang application '~ts'. ~ts",
                                         [File, App, hint()]),
                              exit_or_halt(Reason, StartFlag);
                          _ ->
			      ok
                      end
              end, Mods);
        _ ->
            %% No modules? This is strange
            ok
    end.

check_apps() ->
    spawn(
      fun() ->
	      Apps = [ejabberd |
		      [App || {App, _, _} <- application:which_applications(),
			      App /= ejabberd, App /= hex]],
	      ?DEBUG("Checking consistency of applications: ~ts",
		     [misc:join_atoms(Apps, <<", ">>)]),
	      misc:peach(
		fun(App) ->
			check_app_modules(App, true)
		end, Apps),
	      ?DEBUG("All applications are intact", []),
	      lists:foreach(fun erlang:garbage_collect/1, processes())
      end).

-spec exit_or_halt(iodata(), boolean()) -> no_return().
exit_or_halt(Reason, StartFlag) ->
    ?CRITICAL_MSG(Reason, []),
    if StartFlag ->
            %% Wait for the critical message is written in the console/log
            halt();
       true ->
            erlang:error(application_start_failed)
    end.

get_module_file(App, Mod) ->
    BaseName = atom_to_list(Mod),
    case code:lib_dir(App) of
        {error, _} ->
            BaseName;
        Dir ->
            filename:join([Dir, "ebin", BaseName ++ ".beam"])
    end.

module_name([Dir, _, <<H,_/binary>> | _] = Mod) when H >= 65, H =< 90 ->
    Module = str:join([elixir_name(M) || M<-tl(Mod)], <<>>),
    Prefix = case elixir_name(Dir) of
	<<"Ejabberd">> -> <<"Elixir.Ejabberd.">>;
	Lib -> <<"Elixir.Ejabberd.", Lib/binary, ".">>
    end,
    misc:binary_to_atom(<<Prefix/binary, Module/binary>>);

module_name([<<"auth">> | T] = Mod) ->
    case hd(T) of
        %% T already starts with "Elixir" if an Elixir module is
        %% loaded with that name, as per `econf:db_type/1`
        <<"Elixir", _/binary>> ->  misc:binary_to_atom(hd(T));
        _ -> module_name([<<"ejabberd">>] ++ Mod)
    end;

module_name([<<"ejabberd">> | _] = Mod) ->
    Module = str:join([erlang_name(M) || M<-Mod], $_),
    misc:binary_to_atom(Module);
module_name(Mod) when is_list(Mod) ->
    Module = str:join([erlang_name(M) || M<-tl(Mod)], $_),
    misc:binary_to_atom(Module).

elixir_name(Atom) when is_atom(Atom) ->
    elixir_name(misc:atom_to_binary(Atom));
elixir_name(<<H,T/binary>>) when H >= 65, H =< 90 ->
    <<H, T/binary>>;
elixir_name(<<H,T/binary>>) ->
    <<(H-32), T/binary>>.

erlang_name(Atom) when is_atom(Atom) ->
    misc:atom_to_binary(Atom);
erlang_name(Bin) when is_binary(Bin) ->
    Bin.

format_error({Reason, File}) when is_list(Reason), is_list(File) ->
    Reason ++ ": " ++ File;
format_error(Term) ->
    io_lib:format("~p", [Term]).

hint() ->
    "This usually means that ejabberd or Erlang "
    "was compiled/installed incorrectly.".
